### PR TITLE
cache participants, studies and device entity key ids

### PR DESCRIPTION
**Pull request by anzioka**
PR: scheduled task runs every minute to:
- Cache participants
- Cache device entity key ids
- Cache study entity key ids

### CODE REVIEW

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 1]
```diff
@@ -353,49 +368,10 @@ private UUID getStudyEntityKeyId( UUID organizationId, UUID studyId ) {
...
 
-        } catch ( Exception e ) {
-            logger.error( "failed to get participant EKID: participantId == {}, study = {}, organizationId = {}",
-                    participantId,
-                    studyId,
-                    organizationId,
-                    e );
-            return null;
-        }
+        return participantsById.getOrDefault( participantId, null );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:374]: `.get()` returns `null` by default, so I don't think you need to do `getOrDefault`.

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 2]
```diff
@@ -132,12 +132,15 @@
 
     private final long ENTITY_SETS_REFRESH_INTERVAL = 15 * 60 * 1000; // 15 minutes
 
+    // orgId -> participantId -> EKID
+    private final Map<UUID, Map<String, UUID>> studyParticipants = Maps.newHashMap();
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:136]: This structure basically means that participant ids HAVE to be unique across studies, which doesn’t seem ideal. We've already seen researchers using this pattern... assigning the same participant id to the same human but in different studies.

**Review by UnsungHero97 — CHANGES_REQUESTED** [review]

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java (hunk 2 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:136]: Right!!! Will change.

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 3]
```diff
@@ -131,6 +133,16 @@
     protected static final Logger logger = LoggerFactory.getLogger( ChronicleServiceImpl.class );
 
     private final long ENTITY_SETS_REFRESH_INTERVAL = 15 * 60 * 1000; // 15 minutes
+    private final long STUDY_INFO_REFRESH_INTERVAL  = 60 * 1000; // 1 minute
+
+    // orgId -> studyId -> participantId -> EKID
+    private final Map<UUID, Map<UUID, Map<String, UUID>>> studyParticipantsByOrg = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (makosblade) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:139]: Actually it looks like that .values().stream() is just for logging so these can all stay `HashMap`s

**Review by makosblade — APPROVED** [review]

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java (hunk 1 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java]: `.get()` will throw a NPE if the key doesn't exist unless the map permits null values, which is not the case here

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 4]
```diff
@@ -1324,6 +1244,145 @@ public boolean isKnownParticipant( UUID organizationId, UUID studyId, String par
...
+            orgEntitySets.forEach( ( orgId, templateTypeESIDMap ) -> {
+
+                Map<UUID, UUID> studyIds = Maps
+                        .newHashMap(); // studyEKID -> studyId (for lookup when processing neighbors)
+
+                // entity set ids
+                UUID studiesESID = templateTypeESIDMap.getOrDefault( STUDIES, null );
+                UUID participantsESID = templateTypeESIDMap.getOrDefault( PARTICIPANTS, null );
+                UUID participatedInESID = templateTypeESIDMap.getOrDefault( PARTICIPATED_IN, null );
+                UUID usedByESID = entitySetIdsByOrgId.getOrDefault( CHRONICLE_DATA_COLLECTION, Map.of() )
```
Comment (katherinebernstein) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:1277]: why are different entity sets being loaded in different ways? i.e. 
```
templateTypeESIDMap.getOrDefault( <collectionTypeName>, null );
```
vs
```
entitySetIdsByOrgId.getOrDefault( CHRONICLE_DATA_COLLECTION, Map.of() )
                        .getOrDefault( orgId, Map.of() ).getOrDefault( <collectionTypeName>, null );
```

**Review by katherinebernstein — APPROVED** [review]

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java (hunk 4 above)
Comment (anzioka) [reply src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:1277]: The participants, studies and participatedIn entity sets are in the `CHRONICLE` entity set collection.  perhaps It's cleaner to load all of them the same way.

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 5]
```diff
@@ -1324,6 +1244,145 @@ public boolean isKnownParticipant( UUID organizationId, UUID studyId, String par
...
+
+                    Map<UUID, UUID> mappedEntities = studyEntityKeyIds.getOrDefault( orgId, Maps.newHashMap() );
+                    mappedEntities.put( studyId, studyEKID );
+                    studyEntityKeyIds.put( orgId, mappedEntities );
+                    studyIds.put( studyEKID, studyId );
+                } );
+
+                // get study neighbors constrained by used_by and participated_in associations
+                Set<UUID> edgeESIDS = Sets.newHashSet( participatedInESID, usedByESID )
+                        .stream().filter( Objects::nonNull )
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:1311]: Each operation should be on its own line.

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 6]
```diff
@@ -1324,6 +1244,145 @@ public boolean isKnownParticipant( UUID organizationId, UUID studyId, String par
...
+
+                // process neighbors
+                Map<UUID, Map<String, UUID>> participantsByStudy = Maps
+                        .newHashMap(); // studyId -> participantId -> EKID
+                studyNeighbors.forEach( ( studyEKID, neighbors ) -> {
+                    UUID studyId = studyIds.get( studyEKID );
+
+                    neighbors.forEach( neighbor -> {
+                        if ( neighbor.getAssociationEntitySet().getId().equals( usedByESID ) ) {
+                            String deviceId = getFirstValueOrNull( neighbor.getNeighborDetails().get(), STRING_ID_FQN );
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:1336]: `neighbor.getNeighborDetails()` is optional, so it's possible for `.get()` to actually throw if the value is null.

File: src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java [hunk 7]
```diff
@@ -1324,6 +1244,145 @@ public boolean isKnownParticipant( UUID organizationId, UUID studyId, String par
...
+                            UUID deviceEKID = neighbor.getNeighborId().get();
+
+                            Map<String, UUID> deviceEntities = deviceEntityKeyIds
+                                    .getOrDefault( orgId, Maps.newHashMap() );
+                            deviceEntities.put( deviceId, deviceEKID );
+
+                            deviceEntityKeyIds.put( orgId, deviceEntities );
+                        } else {
+                            // edge: participatedIn
+                            String participantId = getFirstValueOrNull( neighbor.getNeighborDetails().get(),
```
Comment (UnsungHero97) [inline src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java:1349]: Same comment about optionals.

**Review by UnsungHero97 — APPROVED** [review]

**Pull request merged**